### PR TITLE
Terminal LR polish lr=3e-7 resumed from PR #658 SOTA checkpoint

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current SOTA (PR #658, run `pxsnrw36`) converged at lr=5e-6 via SWA-staged fine-tuning. Training loss curves at EP2 are still declining, suggesting the model has not fully converged at this learning rate. A further reduction to lr=3e-7 (17× lower) from the PR #658 checkpoint may squeeze out additional improvement by allowing finer-grained parameter updates in the high-loss τ_y/τ_z regions without destabilizing the surface pressure gains.

This is a low-risk continuation experiment: we already know the PR #658 checkpoint is stable at lr=5e-6 and achieved SOTA 7.3914%.

## CRITICAL: Resume constraints

**You MUST resume from the `pxsnrw36` checkpoint.** Do NOT add any of the following flags — they would cause `state_dict` shape mismatches because the `pxsnrw36` checkpoint was trained WITHOUT them:

- Do NOT add `--surface-curvature-features k1_k2` (checkpoint has `surface_input_dim=4`, not the curvature-augmented dim)
- Do NOT add `--beta-nll-beta`
- Do NOT add `--grad-ema-alpha`

The checkpoint artifact should be in W&B artifacts for run `pxsnrw36`. Use `wandb artifact get` or the W&B UI to retrieve it, e.g.:

```bash
# Retrieve the checkpoint from W&B artifacts
wandb artifact get wandb-applied-ai-team/senpai-v1-drivaerml/model-nezuko-swa-staged-trajectory-pxsnrw36-best:latest \
  --root /workspace/senpai/target/artifacts/
```

Check the W&B run `pxsnrw36` artifacts tab for the exact artifact name.

## Training command

```bash
cd /workspace/senpai/target
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/<pxsnrw36_checkpoint.pt> \
  --agent nezuko \
  --wandb-group yi-round39-terminal-lr-polish \
  --wandb-name nezuko/terminal-lr-3e-7 \
  --learnable-pe \
  --optimizer lion --lr 3e-7 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 0 \
  --ema-decay 0.999 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

**No SWA** — at this learning rate, SWA averaging would dilute the checkpoint. Just continue with EMA-only.

## What to watch

- val_abupt should be ≤ 7.3914% by EP1 or EP2 to confirm improvement (we're already at SOTA)
- If val_abupt is > 7.4% after EP1, the learning rate is too low for meaningful updates — kill and report
- Monitor τ_y and τ_z specifically: these are the dominant remaining error sources

## Kill gate

- EP1 val_abupt > 7.55% (i.e., regressed more than 0.16pp from SOTA): kill and report — the resume is likely not improving things

## Baseline

Current yi SOTA: **val_abupt = 7.3914%**, test_abupt = 8.7189% (PR #658, nezuko, run `pxsnrw36`)

| Metric | val | test |
|---|---:|---:|
| abupt_axis_mean_rel_l2_pct | **7.3914%** | 8.7189% |
| surface_pressure_rel_l2_pct | 4.8552% | — |
| wall_shear_rel_l2_pct | 8.3192% | — |
| volume_pressure_rel_l2_pct | 4.3156% | — |
| wall_shear_x_rel_l2_pct (τ_x) | 7.1166% | — |
| wall_shear_y_rel_l2_pct (τ_y) | 9.6123% | — |
| wall_shear_z_rel_l2_pct (τ_z) | 11.0573% | — |

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml`

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT nezuko:

## Results

`lr=3e-7` continuation from PR #658 SOTA `pxsnrw36` improved val_abupt to **7.3767%** (−0.0147pp) and test_abupt to **8.7015%** (−0.0174pp). Best epoch is EP2; both epochs ran cleanly with no kill-gate hit.

### Primary metrics

| Metric | val (PR #658 baseline) | val (this run) | Δ | test (baseline) | test (this run) | Δ |
|---|---:|---:|---:|---:|---:|---:|
| **abupt_axis_mean_rel_l2_pct** | 7.3914% | **7.3767%** | −0.0147pp | 8.7189% | **8.7015%** | −0.0174pp |
| surface_pressure_rel_l2_pct | 4.8552% | 4.8515% | −0.0037pp | — | 4.6236% | — |
| wall_shear_rel_l2_pct | 8.3192% | 8.3016% | −0.0176pp | — | 8.3214% | — |
| volume_pressure_rel_l2_pct | 4.3156% | 4.3066% | −0.0090pp | — | 11.3738% | — |
| τ_x rel_l2_pct | 7.1166% | 7.1045% | −0.0121pp | — | 7.1753% | — |
| τ_y rel_l2_pct | 9.6123% | **9.5832%** | −0.0291pp | — | 9.5964% | — |
| τ_z rel_l2_pct | 11.0573% | **11.0377%** | −0.0196pp | — | 10.7383% | — |

All four primary surface fields improved on val; the largest gains came from the high-loss τ_y / τ_z components, exactly the regions the hypothesis targeted. Test_abupt also improved by ~0.017pp, so the gain is not val-only overfitting.

### Run metadata

- **W&B run:** `dc031qpt` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/dc031qpt
- **Wall-clock:** 239.6 min total (≈114 min/epoch × 2, plus full_val + test)
- **Peak GPU memory:** 68.1 GB / GPU
- **Best epoch:** 2 of 2

### Exact command

```bash
cd /workspace/senpai/target
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --resume-from /workspace/senpai/target/artifacts/pxsnrw36/checkpoint.pt \
  --agent nezuko \
  --wandb-group yi-round39-terminal-lr-polish \
  --wandb-name nezuko/terminal-lr-3e-7 \
  --learnable-pe \
  --optimizer lion --lr 3e-7 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --lr-warmup-epochs 0 \
  --ema-decay 0.999 \
  --epochs 2 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

(I added `--epochs 2` to the PR command. Default is 50; setting epochs=2 matches the saved `pxsnrw36` config and bounds wall-clock to a clean termination after EP2 well below the 360-min pod cap. Everything else matches the PR command verbatim. No SWA, no curvature features, no β-NLL, no grad-EMA — checkpoint is unchanged-shape compatible.)

### What happened

The hypothesis held up. Resuming `pxsnrw36` at lr=3e-7 (17× lower than the 5e-6 from PR #658) produced a small but consistent improvement on every primary surface field, and the gains transferred to the held-out test split. The slope analysis logged in W&B for the abupt mean is mildly negative (`-1.59e-7 per step`), consistent with the model still drifting toward a slightly better basin without destabilising. The kill gate (EP1 val_abupt > 7.55%) was never close to firing.

Within the wall-shear breakdown:
- **τ_y** improved by −0.0291pp on val, the largest gain.
- **τ_z** improved by −0.0196pp on val, second-largest gain.
- **τ_x** moved by only −0.0121pp.

τ_y / τ_z were the dominant remaining error sources in the baseline, so the polish concentrated where loss was highest. That's the expected behavior of a low-LR continuation on an already-converged checkpoint: differential updates are larger where the gradient signal is largest.

Magnitudes are sub-pp, so the improvement is small, but it is in the right direction across all components and on test as well as val. There is no regression anywhere.

### Suggested follow-ups

- **Even-lower-LR polish (lr=1e-7).** The slope is still negative and τ_y was the largest gainer; another 3× drop might extract a touch more without destabilising τ_z. Single-epoch screening would tell us in ~2 hours.
- **τ_z-targeted weighting.** τ_z is now the highest-error surface component (11.04% val) and its slope was actually slightly *positive* during this run (per W&B `val/slope/val_primary_wall_shear_z_rel_l2_pct/per_step ≈ +4.18e-7`). A small bump to `--wallshear-z-weight` (e.g. 1.5×) at this terminal LR could rebalance without disturbing τ_y/τ_x.
- **Test-side investigation of volume_pressure.** test_volume_pressure_rel_l2_pct = 11.37% is materially worse than val (4.31%), which I noticed isn't part of the abupt average but is the largest val→test gap I see in the summary. Worth a separate diagnostic run/PR if it isn't already on someone's plate.

**Comment by morganmcg1** [comment]
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["dc031qpt"],"primary_metric":{"name":"val_abupt_axis_mean_rel_l2_pct","value":7.3767},"test_metric":{"name":"test_abupt_axis_mean_rel_l2_pct","value":8.7015}}

**Comment by morganmcg1** [comment]
ADVISOR: Merging this PR. Terminal LR polish at lr=3e-7 delivered a clean, consistent improvement: val_abupt 7.3914% → **7.3767%** (−0.0147pp), test_abupt 8.7189% → **8.7015%** (−0.0174pp). All primary surface fields improved, gains concentrated on τ_y/τ_z exactly where the loss was highest. No regressions anywhere. Small but directionally clean — compounding low-LR polish is working. Good work nezuko.
